File: magellan-common/src/main/java/com/wealthfront/magellan/transitions/CrossfadeTransition.java
package com.wealthfront.magellan.transitions;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.view.View;

import com.wealthfront.magellan.Direction;
import com.wealthfront.magellan.NavigationType;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

public class CrossfadeTransition implements Transition {

  @Override
  public void animate(
      final View from, final View to, NavigationType navType, Direction direction, final Callback callback) {
    to.setAlpha(0f);
    from.setAlpha(1f);
    to.setVisibility(VISIBLE);
    from.setVisibility(VISIBLE);

    from.animate().alpha(0f).setListener(new AnimatorListenerAdapter() {
      @Override
      public void onAnimationEnd(Animator animation) {
        from.setVisibility(GONE);
        to.animate().alpha(1f).setListener(new AnimatorListenerAdapter() {
          @Override
          public void onAnimationEnd(Animator animation) {
            callback.onAnimationEnd();
          }
        }).start();
      }
    }).start();
  }

}

File: magellan-kotlin/src/main/java/com/wealthfront/magellan/kotlin/transitions/CrossfadeTransition.kt
package com.wealthfront.magellan.kotlin.transitions

import com.wealthfront.magellan.kotlin.NavigationType
import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import com.wealthfront.magellan.kotlin.Direction


class CrossfadeTransition : Transition {

  override fun animate(from: View, to: View, navType: NavigationType, direction: Direction, callback: Transition.Callback) {
    to.alpha = 0f
    from.alpha = 1f
    to.visibility = VISIBLE
    from.visibility = VISIBLE

    from.animate().alpha(0f).setListener(object : AnimatorListenerAdapter() {
      override fun onAnimationEnd(animation: Animator) {
        from.visibility = GONE
        to.animate().alpha(1f).setListener(object : AnimatorListenerAdapter() {
          override fun onAnimationEnd(animation: Animator) {
            callback.onAnimationEnd()
          }
        }).start()
      }
    }).start()
  }

}